### ISSUE TITLE
Mac OS X: Silent failure when pkgutil setting isn't removed

### ISSUE THREAD

**Issue by noahgibbs** [jtimberman/vagrant-cookbook#22]
When somebody has simply removed the Vagrant directory themselves but left the pkgutil setting in place, Vagrant won't be installed but there will be no error. It'd be nice if the cookbook checked for this condition in its most basic form (no /opt/vagrant directory or local equivalent.)

This has happened to me and at least one of my users, so it's not a pure theoretical concern :-)

**Comment by jtimberman** [comment]
This is more in scope for a pkgutil provider. That's currently done in the dmg cookbook. Perhaps adding an uninstall action to the dmg_package resource would be a good first step?

Longer term I think we should have a pkgutil resource in core Chef.

**Comment by jtimberman** [comment]
Ref https://github.com/opscode-cookbooks/dmg/pull/22

**Comment by noahgibbs** [comment]
Okay. So barring changes in core Chef, is the answer "have cookbooks that use vagrant-cookbook manually check for this condition for MacOS and then use a dmg block to uninstall?"

That's possible, but doesn't feel particularly clean to me. Basically every user of vagrant-cookbook on Mac should be doing that.

**Comment by jtimberman** [comment]
What is the situation when this is necessary? If you have users that are deleting their vagrant directories, I'd find out why and perhaps try to educate them that it's a bad idea that can lead to maintenance problems? Otherwise, package upgrades should work in place, as Vagrant doesn't have user-side installation in their all-in-one directory. 

As for this: 

> have cookbooks that use vagrant-cookbook manually check for this condition for MacOS and then use a dmg block to uninstall?

Maybe? I don't know what your cookbooks look like, or what users are doing with them on their OS X systems. You can certainly automate around the problem using Chef, but I really would try to get to the heart of what the problem is and try to solve that.

FWIW, there's also homebrew-cask, which I would strongly consider for installing Vagrant in a way that lets you worry less about upgrade or uninstall scenarios.

```
% brew cask info vagrant
vagrant: 1.7.2
Vagrant
https://www.vagrantup.com/
/opt/homebrew-cask/Caskroom/vagrant/1.7.2 (6 files, 79M)
https://github.com/caskroom/homebrew-cask/blob/master/Casks/vagrant.rb
==> Contents
  Vagrant.pkg (pkg)
```

The homebrew cookbook has a `homebrew_cask` resource that has `:install` and `:uninstall` actions.

**Comment by noahgibbs** [comment]
My use case is here: "https://github.com/noahgibbs/madscience". It's an attempt to install a deterministic Chef-based stack of tools to do Ruby installation. I am being reminded why Docker is kicking everybody else's ass in this space -- it's nearly impossible to install a fixed, known, tested stack of tools, including things like a very specific Vagrant version that you've actually tested.

Deleting programs is often considered a good way to uninstall them. In the current situation, if a user tries that with Vagrant, it appears to work. But then future Vagrant installs silently fail.

If "educate users" is the solution then I have to tell anybody who ever uninstalled Vagrant that way (including one of my first four test users) that they have to not do that. They often did it well before I ever talked to them. Vagrant's uninstall instructions, the real ones, take some work to find. Somebody just trying to uninstall Vagrant is very unlikely to bother. I've done that. I've had random users do that before finding my software.

Currently, my cookbook detects that situation and reminds them of what they may or may not remember doing at some point in the past. So I have the higher levels not written as "install Vagrant", but "attempt to install Vagrant, see if it fails, and then check if we're on a Mac where there's a standard, known case where Vagrant screws that up." That sounds like it wants to be at the "vagrant cookbook" level of abstraction to me -- part of the "install vagrant" level rather than handled as a normal but undocumented failure case by anybody who uses it.

User education is not a great solution to this. Silent failure is also not a great solution to this.

It would be nice if Vagrant didn't set things up this way, of course.

So: I solved this with user education the twice now that I've seen it. Which means I've solved it by walking through the person's bug reports and eventually saying, "oh hey, is it possible that you did this reasonable-sounding thing at some point in the past? That totally breaks everything until you do this cryptic series of steps that are hard to Google for."

Then I added the check to my cookbook -- don't just install Vagrant, check that it's in reasonable shape afterward and detect this weird Mac-specific use case where it fails.

Maybe that's the right answer. Very often, it seems like Chef cookbooks are written in such a way that they screw up if you don't install them on a new clean system. But Vagrant in particular seems extra-likely to be installed on somebody's dev box, not on a new clean system.

That's certainly what I'm trying to do.
